### ISSUE TITLE
Bitmap blit blending algorithm (bitmapBlit.frag) is not 100% accurate for color values.

### ISSUE THREAD

**Issue by Ancurio** [Ancurio/mkxp#14]
According to my personal tests, the resulting alpha value is correct for all inputs (srcAlpha, dstAlpha, bltAlpha*), however, the color value is just an approximation as I have yet to successfully reverse-engineer the algorithm behind it. The approximation is written to favor text display (with eg. manual shadows, or at half opacity) as much as possible, but many visible irregularities remain and the code itself is quite ugly. Note that in case of blitting with bltAlpha = 1.0 on a cleared surface (the vanilla way text is displayed in RMXP games), no blending is necessary and the source is directly copied into the destination (ie. a literal blit).

The color value has 4 inputs: srcColor, dstColor, srcAlpha*bltAlpha and dstAlpha, The results get especially confusing once the alpha values start deviating from 1.0.

(*bltAlpha is the 'opacity' parameter passed into the blt functions, or the alpha value of the font used when drawing text).

**Comment by fantasydr** [comment]
> But seeing that I couldn't find a shred of useful information in MS' GDI docs, it makes sense that RMXP must be rolling its own blending code.

@Ancurio @cremno
Sorry for replying to this issue after so many years. I'm not sure if you ever found the answer, but here it is:

TL;DR: This blending method is called Porter-Duff "source over" blending, originating from an old paper:
[Porter-Duff (1984)](https://keithp.com/~keithp/porterduff/p253-porter.pdf)


And the long version...  

Back in the summer of 2013, I was also trying to make a Linux version of RMXP, and I ran into a similar issue: the blt method in RMXP is not a simple alpha-blend. So, I dug into its machine code and finally found the real implementation. It is written in MMX asm and involves two pre-defined buffers. 

I tried to understand the algorithm and re-implemente in c++, but I still couldn't figure out where these two buffers came from. I knew it must be an optimization for some function, but I couldn't figure out exactly which one. Eventually, I didn't finish that project, and it was buried in my codebase.

Lately, I noticed that AI models are very good at analyzing code, so I let one analyze the re-implementation code. Claude Opus 4.6 did a very good job and gave me a detailed explanation about it.

I put my code snippets and the explanation into a repo here, in case anyone in the future is interested:
https://github.com/fantasydr/rmxp-bitmap-blt

BTW, since RMXP used quantized blending weights instead of the exact mathematical formula, it generates slightly different results. I think the shader code in this project is perfect as it is, because it faithfully reflects the original Porter-Duff blending method.

**Comment by Ancurio** [comment]
@fantasydr wild that you found the original algorithm (and paper too). Unfortunately decompiling the rmxp binaries is something I never intended to do in order for everything to truly be white room implemented, but it's helpful if others do it and then document the behavior; that's what I had to do for rgssad support as well.

In any case, thanks for documenting this!

**Comment by fantasydr** [comment]
> Unfortunately decompiling the rmxp binaries is something I never intended to do in order for everything to truly be white room implemented.

I didn't publish the project back then because I was worried it might contain a proprietary algorithm from RMXP. Now that I know it's simply based on a published paper using quantized weights as an optimization, I feel it's safe to share the code snippet.

Anyway, thank you for your great work in helping to port and preserve RPG Maker games.

**Comment by cremno** [comment]
@fantasydr: https://github.com/ain1084/nxlib/blob/1a7658c22f2596fbb7eed6965e11b7ab6d590840/src/nxdraw/NxDynamicDraw32.cpp :)

**Comment by fantasydr** [comment]
> [@fantasydr](https://github.com/fantasydr): https://github.com/ain1084/nxlib/blob/1a7658c22f2596fbb7eed6965e11b7ab6d590840/src/nxdraw/NxDynamicDraw32.cpp :)

Wow, so this is the lib used by RMXP! I didn't know the source code was available. Back then when I dug into the machine code, I found a dynamic pipeline for the CPU that generated shader-like drawing code at runtime. It was super cool to analyze, and it made me believe the RMXP author is a hardcore graphics programmer.

Thanks a lot for sharing this! That solves a lot of puzzles for me.
